File: VB/GridMultiTouchProject/GridMultiTouchBehavior.vb
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Grid

Namespace GridMultiTouchProject

    Friend Class GridMultiTouchBehavior
        Inherits Behavior(Of GridControl)

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            owner = AssociatedObject
            AddHandler owner.MouseRightButtonUp, AddressOf owner_MouseRightButtonUp
            AddHandler owner.MouseLeftButtonDown, AddressOf owner_MouseLeftButtonDown
            _isCanMultiSelect = False
        End Sub

        Private _startRowSelect As Integer

        Private _finishRowSelect As Integer

        Private owner As GridControl

        Private _isMultiSelection As Boolean

        Private _isCanMultiSelect As Boolean

        Private tableView1 As TableView

        Private Sub SelectRows()
            owner.UnselectAll()
            owner.SelectRange(_startRowSelect, _finishRowSelect)
        End Sub

        Private Sub tableView1_ShowingEditor(ByVal sender As Object, ByVal e As ShowingEditorEventArgs)
            If _isMultiSelection Then e.Cancel = True
        End Sub

        Private Sub owner_MouseLeftButtonDown(ByVal sender As Object, ByVal e As System.Windows.Input.MouseButtonEventArgs)
            If _isMultiSelection Then _isMultiSelection = False
        End Sub

        Private Sub owner_MouseRightButtonUp(ByVal sender As Object, ByVal e As System.Windows.Input.MouseButtonEventArgs)
            If tableView1 Is Nothing Then
                tableView1 = TryCast(owner.View, TableView)
                If tableView1 IsNot Nothing Then
                    AddHandler tableView1.ShowingEditor, AddressOf tableView1_ShowingEditor
                    _isCanMultiSelect = True
                End If
            End If

            If _isCanMultiSelect Then
                If Not _isMultiSelection Then
                    _isMultiSelection = True
                    _startRowSelect = tableView1.FocusedRowHandle
                    _finishRowSelect = tableView1.FocusedRowHandle
                    tableView1.CancelRowEdit()
                    SelectRows()
                Else
                    _finishRowSelect = tableView1.FocusedRowHandle
                    SelectRows()
                End If
            End If
        End Sub
    End Class
End Namespace

File: CS/GridMultiTouchProject/GridMultiTouchBehavior.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Interactivity;
using DevExpress.Xpf.Grid;

namespace GridMultiTouchProject {
    class GridMultiTouchBehavior : Behavior<GridControl> {
        protected override void OnAttached() {
            base.OnAttached();
            owner = this.AssociatedObject;
            owner.MouseRightButtonUp += owner_MouseRightButtonUp;
            owner.MouseLeftButtonDown += owner_MouseLeftButtonDown;
            _isCanMultiSelect = false;
        }
        int _startRowSelect;
        int _finishRowSelect;
        GridControl owner;
        bool _isMultiSelection;
        bool _isCanMultiSelect;
        TableView tableView1;

        void SelectRows() {
            this.tableView1.ClearSelection();
            this.tableView1.SelectRange(_startRowSelect, _finishRowSelect);
        }

        void tableView1_ShowingEditor(object sender, ShowingEditorEventArgs e) {
            if(_isMultiSelection)
                e.Cancel = true;
        }
        void owner_MouseLeftButtonDown(object sender, System.Windows.Input.MouseButtonEventArgs e) {
            if(_isMultiSelection)
                _isMultiSelection = false;
        }
        void owner_MouseRightButtonUp(object sender, System.Windows.Input.MouseButtonEventArgs e) {
            if(tableView1 == null) {
                tableView1 = owner.View as TableView;
                if(tableView1 != null) {
                    tableView1.ShowingEditor += tableView1_ShowingEditor;
                    _isCanMultiSelect = true;
                }
            }
            if(_isCanMultiSelect) {
                if(!_isMultiSelection) {
                    _isMultiSelection = true;
                    _startRowSelect = this.tableView1.FocusedRowHandle;
                    _finishRowSelect = this.tableView1.FocusedRowHandle;
                    this.tableView1.CancelRowEdit();
                    SelectRows();
                }
                else {
                    _finishRowSelect = this.tableView1.FocusedRowHandle;
                    SelectRows();
                }
            }
        }
    }
}
